katcp/resource.py

```diff
@@ -894,9 +892,7 @@ def handle_update(sensor, reading):
             # This handler is called whenever a sensor update is received
             try:
                 assert sensor is self
-                cond_matched = condition_test(reading)
-                status_matched = reading.status == status or status is None
```

Comment:
Not true on both counts...

`condition_test` takes an entire reading and is free to check value, status and/or timestamps. You are maybe confused with the case where `condition_or_value` is an actual value as opposed to a generic function :-)

The explicit status check was introduced by yours truly in the katuilib days already. The evidence:

> r12268 | schwardt | 2012-01-24 09:59:41 +0200 (Tue, 24 Jan 2012) | 1 line
> 
> (schwardt, simonr) Added the ability to wait until an arbitrary condition is satisfied on a sensor, in addition to the default condition of the sensor value being equal to a desired value. The differing wait implementations in KATClient and Array are now unified with the same parameters and behaviour. The signal implementation in KATClient has been replaced by a time-based implementation, which is shorter and slightly easier to understand. The timeouts were made more accurate. Both versions now throw exceptions on missing sensors or ones without a strategy, and both respond to a KeyboardInterrupt. The updated signal implementation in KATClient is kept in commented-out form for posterity.
> 

That is why one writes decent commit messages... Unfortunately it was not decent enough to remind me of the need for an explicit status check :-)

I think my thinking was that you always only want to wait on a 'nominal' sensor, and if that is the case you might as well make it explicit instead of forcing a lambda-type check all the time.

Then, in the translation to katcp, the default became `status=None`, so that whole thing went out the door. But nothing much went wrong, so I guess my need for 'nominal'-ness has been misplaced :-) It sort of makes sense, as warn and error states do not invalidate the actual value of the sensor and should still be wait()-able.